File: leetcode/editor/cn/108_将有序数组转换为二叉搜索树_test.py
import unittest
from typing import List, Optional

from model.tree_node import TreeNode


class TestSolution(unittest.TestCase):

    def setUp(self) -> None:
        self.solution = Solution()

    def test_givenNormal_when_thenSuccess(self):
        self.assertTrue(True, "ShouldBeTrue")


# leetcode submit region begin(Prohibit modification and deletion)
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sortedArrayToBST(self, nums: List[int]) -> Optional[TreeNode]:
        if len(nums) == 0:
            return None

        # TODO len(nums)/2 result is float
        # pivot: int = int(len(nums) / 2)
        pivot: int = len(nums) // 2
        return TreeNode(nums[pivot],
                        self.sortedArrayToBST(nums[:pivot]),
                        self.sortedArrayToBST(nums[pivot + 1:])
                        )

        pass


# leetcode submit region end(Prohibit modification and deletion)


if __name__ == '__main__':
    unittest.main()

File: leetcode/editor/cn/108_将有序数组转换为二叉搜索树_test.go
package leetcode

import (
	"github.com/stretchr/testify/assert"
	"testing"
)

func Test_givenNormal_when108_thenSuccess(t *testing.T) {
	root := sortedArrayToBST([]int{-10, -3, 0, 5, 9})
	nodes := encodeTreeNode(root)
	assert.Equal(t, []any{0, -3, 9, -10, nil, 5}, nodes)
}

//leetcode submit region begin(Prohibit modification and deletion)
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func sortedArrayToBST(nums []int) *TreeNode {
	if len(nums) == 0 {
		return nil
	}
	pivot := len(nums) / 2
	left := sortedArrayToBST(nums[0:pivot])
	right := sortedArrayToBST(nums[pivot+1:])
	return &TreeNode{Val: nums[pivot], Left: left, Right: right}
}

//leetcode submit region end(Prohibit modification and deletion)
